test_util/provide_aws.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+import logging
+import os
+
+import boto3
+import retrying
+
+LOGGING_FORMAT = '[%(asctime)s|%(name)s|%(levelname)s]: %(message)s'
+logging.basicConfig(format=LOGGING_FORMAT, level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+# Allows adjusting the timeout in CI
+AWS_WAITING_TIMEOUT = os.getenv('AWS_WAITING_TIMEOUT', 60)
+
+
+class BotoWrapper():
+    def __init__(self, region, aws_access_key_id, aws_secret_access_key):
+        self.region = region
+        self.session = boto3.session.Session(
+                aws_access_key_id=aws_access_key_id, aws_secret_access_key=aws_secret_access_key)
+
+    def client(self, name):
+        return self.session.client(service_name=name, region_name=self.region)
+
+    def resource(self, name):
+        return self.session.resource(service_name=name, region_name=self.region)
+
+    def create_key_pair(self, key_name):
+        """Retruns private key of newly generated pair
+        """
+        key = self.client('ec2').create_key_pair(KeyName=key_name)
+        return key['KeyMaterial']
+
+    def delete_key_pair(self, key_name):
+        self.client('ec2').delete_key_pair(KeyName=key_name)
+
+    def create_stack(self, name, template_url, user_parameters):
+        """Returns boto stack object
+        """
+        logging.info('Requesting AWS CloudFormation...')
+        cf_parameters = []
+        for k, v in user_parameters.items():
+            cf_parameters.append({'ParameterKey': k, 'ParameterValue': v})
+        return self.resource('cloudformation').create_stack(
+                StackName=name,
+                TemplateURL=template_url,
+                DisableRollback=True,
+                TimeoutInMinutes=AWS_WAITING_TIMEOUT,
+                Capabilities=['CAPABILITY_IAM'],
+                Parameters=cf_parameters)
+
+
+class CFStack():
+    def __init__(self, stack_name, boto_wrapper):
+        self.stack_name = stack_name
+        self.boto_wrapper = boto_wrapper
+
+    def attach(self):
+        self.stack = self.boto_wrapper.resource('cloudformation').Stack(self.stack_name)
+
+    def wait_for_stack_creation(self):
+        self.wait_for_transition('CREATE_IN_PROGRESS', 'CREATE_COMPLETE')
+
+    def wait_for_stack_deletion(self):
+        self.wait_for_transition('DELETE_IN_PROGRESS', 'DELETE_COMPLETE')
+
+    def wait_for_transition(self, start_status, end_status):
+        final_status, status_reason = self.wait_for_status_change(start_status)
+        assert end_status == final_status, "Expected status {} but found {}. Reason: {}".format(
+                end_status, final_status, status_reason)
+
+    @retrying.retry(wait_fixed=5000, stop_max_delay=AWS_WAITING_TIMEOUT*60*1000,
+                    retry_on_result=lambda res: res is False,
+                    retry_on_exception=lambda ex: False)
+    def wait_for_status_change(self, status):
```

Comment:
Sure cleans up everything pretty well except it doesn't allow adjusting the timeout :( That being said, I was just going to use an hour timeout anyway